Prepare the RAND code for moving into the FIPS module

mattcaswell opened the pull request:
This is mostly about changing the way the RAND code initialises itself so that it is based on OPENSSL_CTX instead.

(Needed in order to move the RAND code into the FIPS provider - see #9035)

levitte commented on crypto/init.c:
```diff
@@ -855,5 +852,6 @@ void OPENSSL_fork_parent(void)
 void OPENSSL_fork_child(void)
 {
     rand_fork();
+    /* TODO(3.0): Inform all providers about a fork event */
```
This could be implemented using a provider wide set_params...

levitte commented:
At a first quick look through, this looks ok, but I frankly don't know this code enough.  @mspncp, would you mind having a look?

levitte commented:
Rebase needed for CIs

mattcaswell commented:
Rebased as requested.

mspncp commented:
> @mspncp, would you mind having a look?

I started taking a look and immediately my gut instinct tells me that it's not happy about this proliferation of `underscore_ex` functions introduced for passing `OPENSSL_CTX *` pointers around. In particular, extending _public_ API calls like `RAND_bytes` to `RAND_bytes_ex` feels wrong to me. But I need some time to figure out what precisely disturbs me and whether the problem could be addressed differently...

paulidale commented:
`_ctx` ?

mspncp commented:
> `_ctx` ?

Might be an improvement over the "Microsoft disease" of adding exes to API functions. But I think the problem lies deeper...

paulidale commented:
> > `_ctx` ?
> 
> Might be an improvement over the "Microsoft disease" of adding exes to API functions. But I think the problem lies deeper...

I'm inclined to agree.  The question is what alternative is there?

A per thread context might be workable but represents a significant departure from where things were headed.

mspncp commented:
> I'm inclined to agree. The question is what alternative is there?
> 
> A per thread context might be workable ...

My thoughts go in this direction.  However, it will still be a _library_ context not a _thread_ context. Only the _pointer to the current library context_ will be implemented as thread local variable and there needs to be a simple and consistent way to activate a new context at library boundaries. (These activations need to be stacked.) I'll try to figure out more details, but I first need to understand the current state of the implementation better.

> ... but represents a significant departure from where things were headed.

Well, it's not too late yet to change things:
```
No matter how far down the wrong road you've gone, turn back.
                                             Turkish proverb
[You'll know whether the road is wrong if you carefully listen
 to your conscience.]
```
_(taken from @t8m's email footer)_

mattcaswell commented:
For most APIs (such as EVP) its not too much of a problem. There are plenty of existing FOO_CTX objects that get passed to the various function calls, and therefore you only need a relatively small number of additional functions to associate those FOO_CTX objects with an OPENSSL_CTX - and then suddenly all those functions have the required access to OPENSSL_CTX.

The RAND API is a little different because there is no RAND_CTX concept - so there are more functions that need to change to add support for it. Personally I don't see this as too much of a problem.

I'm not particularly fond of the "_ex" naming scheme, so I'm certainly open to alternative suggestions.

mattcaswell commented:
Rebased again due to yet more conflicts with master

mspncp commented:
> For most APIs (such as EVP) its not too much of a problem.

I see. That makes things a lot simpler and will (hopefully) make thread-local implicit library contexts dispensable. 

> The RAND API is a little different because there is no RAND_CTX concept - so there are more functions that need to change to add support for it.

If this is your only problem, then the solution is easy:  Just use `RAND_DRBG_bytes()` instead of `RAND_bytes()`: contrary to the legacy `RAND` API, the `RAND_DRBG` API does have a context, it is the `RAND_DRBG` struct itself (a.k.a `rand_drbg_st`):

https://github.com/openssl/openssl/blob/ff8029c189786196c010292b647f3bca6d51e91d/include/openssl/ossl_typ.h#L124

The `RAND_DRBG` structure is opaque and it is easy to add a backward pointer from every `RAND_DRBG` struct to the `OPENSSL_CTX` struct which owns it.

https://github.com/openssl/openssl/blob/8094a6945873f492fe40c88b966b86629bc6c6d7/crypto/rand/rand_lcl.h#L193-L197

_Note: In the times of the Fips Object Module 2.0, the `RAND_DRBG` struct used to be called a "drbg context"  ..._


https://github.com/openssl/openssl/blob/d11e6a4410d5aae70ad545d085d344644c33d9cc/fips/rand/fips_rand.h#L76


_... but this terminology was dropped by me in favor of the newer "OpenSSL class" pattern (`CLASS_NAME_method_name(CLASS_NAME *obj, ...)`)._

mspncp commented:
Providers should stick to using the `RAND_DRBG` api directly and the legacy `RAND` api should remain for application use only. 

Then the only thing which remains to do is to redirect the RAND methods to the FIPS provider DRBGs when FIPS mode is enabled.

mattcaswell commented:
> Providers should stick to using the RAND_DRBG api directly and the legacy RAND api should remain for application use only.

Hmm. Right. That does make sense. I'll look into that.

beldmit commented:
It's worth mentioning that in Russia we must (and even MUST) to provide our own certified RNGs. Currently we are able to do it via engines, and we need this possibility via providers.

mattcaswell commented:
> It's worth mentioning that in Russia we must (and even MUST) to provide our own certified RNGs. Currently we are able to do it via engines, and we need this possibility via providers.

That is the plan (although you don't get it through this PR).

mspncp commented:
**Side note:** The `OPENSSL_CTX` typedef-struct fits nicely into the new OpenSSL "class" pattern. Having this in mind, the awkward `underscore_ex` functions

```
RAND_DRBG *RAND_DRBG_get0_master_ex(OPENSSL_CTX *ctx);
RAND_DRBG *RAND_DRBG_get0_public_ex(OPENSSL_CTX *ctx);
RAND_DRBG *RAND_DRBG_get0_private_ex(OPENSSL_CTX *ctx);
```

naturally turn into accessors for the `OPENSSL_CTX` class:

```
RAND_DRBG *OPENSSL_CTX_get_master_drbg(OPENSSL_CTX *ctx);
RAND_DRBG *OPENSSL_CTX_get_public_drbg(OPENSSL_CTX *ctx);
RAND_DRBG *OPENSSL_CTX_get_private_drbg(OPENSSL_CTX *ctx);
```

(The zero in `get0` is not necessary IMO, the `get` could even be dropped entirely. Feel free to choose your favorite variation of the suggested names.)

mspncp commented:
**Side note 2**: Most of the modern style OpenSSL "classes" follow the `OSSL_CLASSNAME` naming pattern, i.e. they use `OSSL_` as the common prefix.

For that reason (and for brevity), I would suggest to rename `OPENSSL_CTX` to `OSSL_CTX`  (and the lowercase versions accordingly). If you agree with me, I volunteer to carry out the straightforward renaming task.

richsalz commented:
> I'm not particularly fond of the "_ex" naming scheme, so I'm certainly open to alternative suggestions.

It is what is used throughout OpenSSL.  Please don't add a new convention; the API is hard enough to understand already.

levitte commented:
> > I'm not particularly fond of the "_ex" naming scheme, so I'm certainly open to alternative suggestions.
> 
> It is what is used throughout OpenSSL. Please don't add a new convention; the API is hard enough to understand already.

side note: I've always wondered what will happen when we need to extend a function that already has `_ex` at the end...

levitte commented:
> **Side note 2**: Most of the modern style OpenSSL "classes" follow the `OSSL_CLASSNAME` naming pattern, i.e. they use `OSSL_` as the common prefix.

`OSSL_` and `OPENSSL_` are both acceptable.

mspncp commented:
> `OSSL_` and `OPENSSL_` are both acceptable.

@levitte what is the correct interpretation of your statement?

1. You have no preference and don't mind if I change the name from `OPENSSL_CTX` to `OSSL_CTX`.
2. You would prefer if I'd leave the names as they are, since both are acceptable.

mspncp commented:
> > I'm not particularly fond of the "_ex" naming scheme, so I'm certainly open to alternative suggestions.
> 
> It is what is used throughout OpenSSL. Please don't add a new convention; the API is hard enough to understand already.

I don't intend to create new conventions, but I'd rather avoid adding meaningless `_ex` suffixes which IMHO should be used only as last resort if you run out of ideas for naming your functions properly ;-). See https://github.com/openssl/openssl/pull/9039#issuecomment-498627564 for an example how it could be done better.

levitte commented:
I'm not strictly opposed to `OSSL_CTX` over `OPENSSL_CTX`, but I don't see that as a necessity either.  In other words, such a name change is very low on my priority list.

mattcaswell commented:
> (The zero in get0 is not necessary IMO, the get could even be dropped entirely. Feel free to choose your favorite variation of the suggested names.)

Why do you think it is unnecessary? It seems to fit the intended meaning here?

mspncp commented:
> Why do you think it is unnecessary? It seems to fit the intended meaning here?

Unnecessary only in the sense that DRBG's are never up-rev'ed, so there is no need to distinguish between those two cases. But if you prefer the `get0` because you think it is more explicit, that's ok for me.

mattcaswell commented:
To me it implies more than just whether an object has been up-ref'd or not. It implies ownership. With a get0 you should not free the returned object explicitly. With a get1 you should. That applies whether or not the object in question is ref counted or not.

So, I prefer to keep it because it makes the intended memory semantics clearer.

levitte commented:
Quite frankly, the get0/get1 and set0/set1 naming scheme isn't very intuitive.  We who have been part of the project for a while know what they mean (and still get it wrong at times), but for someone new to the mind-set?  It's kinda horrible, really...

richsalz commented:
I agree it's not great but don't change it, fix it by adding more docs and cross-referencing all manpages with a [sg]et[01] API to point to the new man7 page.

mspncp commented:
Once you get accustomed to it, it is a help. And it is an improvement, because before this convention was introduced, there were far too many other functions introduced with unclear memory semantics. So I agree with @mattcaswell and @richsalz that we should better stick to it.

mattcaswell commented:
I've removed the parts of this PR that added a lot of the _ex functions and restricted the changes to just the RAND_DRBG API. This will mean that any code brought into the FIPS module will have to be updated to use that API instead...but that's ok.

I also renamed the functions as suggested to `OPENSSL_CTX_get0_public_drbg()` etc.

slontis commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -15,11 +17,20 @@ RAND_DRBG_free
 
  #include <openssl/rand_drbg.h>
 
+ RAND_DRBG *RAND_DRBG_new_ex(OPENSSL_CTX *ctx,
```
RAND_DRBG_ctx_new() maybe?

mattcaswell replied:
Hmm. I'm dubious about adding that as a new convention. What if I later have to extend `FOO_CTX_new()` to add an `OPENSSL_CTX` param? Calling it `FOO_CTX_ctx_new` sounds odd.

slontis commented on crypto/rand/drbg_lib.c:
```diff
@@ -141,6 +147,152 @@ static int is_digest(int type)
+        return NULL;
+
+    if (!CRYPTO_THREAD_init_local(&dgbl->private_drbg, NULL))
+        goto err1;
```
I would prefer seeing the CRYPTO_THREAD_cleanup_local() implementations checked for NULL, rather than doing this.  It looks like basic code :)

mattcaswell replied:
I don't believe this is possible. `&dgbl->private_drbg` here will never be NULL even if it has not been initialised by `CRYPTO_THREAD_init_local`. The contents of `dgbl->private_drbg` might have a zero value because it has been "zalloc'd". But we do not know whether 0 is an invalid value or not for the local platform threading scheme - so it would not be correct for `CRYPTO_THREAD_cleanup_local` to check for 0.

slontis commented on crypto/rand/drbg_lib.c:
```diff
@@ -141,6 +147,152 @@ static int is_digest(int type)
+    return dgbl;
+
+ err3:
+    CRYPTO_THREAD_cleanup_local(&dgbl->public_drbg);
```
drbg_ossl_ctx_free() instead?

mattcaswell replied:
Not possible because I need to have the different entry points for the error handing (err1/err2/err3).

slontis commented on crypto/rand/drbg_lib.c:
```diff
@@ -141,6 +147,152 @@ static int is_digest(int type)
+
+/*
+ * Implements the get_nonce() callback (see RAND_DRBG_set_callbacks())
+ *
```
NIT

slontis commented on crypto/rand/drbg_lib.c:
```diff
@@ -141,6 +147,152 @@ static int is_digest(int type)
+        = openssl_ctx_get_data(drbg->libctx, OPENSSL_CTX_DRBG_NONCE_INDEX,
+                               &drbg_nonce_ossl_ctx_method);
+    struct {
+        void * instance;
```
NIT *instance

slontis commented on test/drbgtest.c:
```diff
@@ -1264,7 +1264,8 @@ static const size_t crngt_num_cases = 6;
 static size_t crngt_case, crngt_idx;
 
-static int crngt_entropy_cb(unsigned char *buf, unsigned char *md,
+static int crngt_entropy_cb(OPENSSL_CTX *ctx, unsigned char *buf,
```
Hmm - is this a backwards compatibility breaking change?

mattcaswell replied:
No. This callback is only exposed because drbgtest.c includes rand_lcl.h

slontis commented on test/drbgtest.c:
```diff
@@ -1288,19 +1289,16 @@ static int test_crngt(int n)
     crngt_case = n % crngt_num_cases;
     crngt_idx = 0;
     crngt_get_entropy = &crngt_entropy_cb;
-    if (!TEST_true(rand_crngt_init()))
```
What is the effect of removing these crngt calls?

mattcaswell replied:
The CRNGT code should be automatically initialised when it attempts to get its data out of the OPENSSL_CTX for the first time. Similarly it is automatically cleaned up when the OPENSSL_CTX gets freed. Therefore there is no need to explicitly initialise and clean-up here.

mattcaswell commented:
Fix up commit pushed addressing the style nits. All other comments responded to above.

mattcaswell commented:
Pushed. Thanks!
